com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/jobs/CleanUpJob.java

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+    }
+
+    private List<Operation> getFinishedOperations(Date expirationTime) {
+        OperationFilter filter = new OperationFilter.Builder().startedBefore(expirationTime).inFinalState().descending().build();
+        return dao.find(filter);
+    }
+
+    private List<Operation> getActiveOperationsInStateError(Date expirationTime) throws SLException {
+        OperationFilter filter = new OperationFilter.Builder().startedBefore(expirationTime).inNonFinalState().descending().build();
+        return operationsHelper.findOperations(filter, Arrays.asList(State.ERROR));
+    }
+
+    private List<String> getProcessIds(List<Operation> operations) {
+        return operations.stream().map(operationInError -> operationInError.getProcessId()).collect(Collectors.toList());
+    }
+
+    private void executeAbortOperationAction(List<String> processIds) {
+        ActivitiAction abortAction = ActivitiActionFactory.getAction("abort", activitiFacade, null);
+        for (String processId : processIds) {
+            abortAction.executeAction(processId);
+        }
+    }
+
+    private void removeProgressMessages(List<String> oldFinishedOperationsIds) {
+        int removedProgressMessages = ProgressMessageService.getInstance().removeAllByProcessIds(oldFinishedOperationsIds);
+        LOGGER.info("Deleted progress messages rows count : " + removedProgressMessages);
+    }
+
+    private void removeProcessLogs(List<String> oldFinishedOperationsIds) {
+        int removedProcessLogs = ProcessLogsService.getInstance().deleteAllByProcessIds(oldFinishedOperationsIds);
+        LOGGER.info("Deleted process logs rows count : " + removedProcessLogs);
+    }
+
+    private Map<String, String> getProcessIdToAppArchiveId() {
+        List<HistoricVariableInstance> appArchiveIds = activitiFacade
+            .getHistoricVariableInstancesByVariableName(Constants.PARAM_APP_ARCHIVE_ID);
+        Map<String, String> processIdToAppArchiveId = new HashMap<>();
+        for (HistoricVariableInstance appArchiveId : appArchiveIds) {
+            processIdToAppArchiveId.put(appArchiveId.getProcessInstanceId(), (String) appArchiveId.getValue());
+        }
+        return processIdToAppArchiveId;
+    }
+
+    private Map<String, List<String>> getSpaceToProcessIds(List<Operation> operations) {
+        return operations.stream()
+            .collect(Collectors.groupingBy(Operation::getSpaceId, Collectors.mapping(Operation::getProcessId, Collectors.toList())));
+    }
+
+    private Map<String, List<String>> getSpaceToFileIds(Map<String, List<String>> spaceToProcessIds,
+        Map<String, String> processIdToAppArchiveId) {
+        if (spaceToProcessIds.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        Map<String, List<String>> spaceToFileIds = new HashMap<String, List<String>>();
+        for (String space : spaceToProcessIds.keySet()) {
+            if (spaceToProcessIds.get(space) == null) {
```

Comment:
Is this needed?

Reply:
Now, all operations should have space. But we don't ensure on DB and code level this, so it is possible to find similar (corrupted) operation. If such operation exists, deletion of mtar archive (fs-storage) or its attributes (LM_SL_PERSISTENCE table) will fail.